docs/webhooks.md

```diff
@@ -0,0 +1,72 @@
+# Webhooks
+
+?> Available `5.2.0`
+
+A variety of webhooks are available to extend the functionality of SQLPad.
+
+To enable a specific webhook, provide a URL for the event you would like to subscribe to. This URL will be sent an HTTP(S) POST with a payload of relevant information when the event occurs.
+
+To ensure a webhook call is valid, `SQLPAD_WEBHOOK_SECRET` may be configured, and validated on every webhook call.
+
+Each webhook will contain the following headers:
+
+- `Content-Type`: `application/json`
+- `SQLPad-Secret`: `SQLPAD_WEBHOOK_SECRET value`
+- `SQLPad-URL`: `http://your-sqlpad-public-url:port/base-url`
+- `SQLPad-Hook-Name`: `hook_name`
+
+The payload sent to the URL varies by event.
+
+## User Created
+
+The user created webhook whenever a user is added via the users API/UI. This hook can be used to automate user invites, sending a message to users inviting them to sign up (if using local user/password authentication) or sign in (if using a single sign on).
+
+**Env**: `SQLPAD_WEBHOOK_USER_CREATED_URL`
+
+**Hook Name**: `user_created`
+
+**Payload Body**:
+
+```js
+{
+  id: '7681aec0-b8c5-4267-b9cb-26f7ff3b3e48',
```

Comment:
what is this id? is this an internal sqlpad user id? if so, that it make sense to  export it?

Reply:
Yeah it is the internal id of the object related to the hook. so for user_created it'd be user id, for query_created query id. 

My thought on sending is to make it available in case someone wants to interact in some way based on an API call or capture it for other purposes? 

I can see the confusion though as to what the id represents. 

Would it make sense to "envelope" the data sent by the hook? For example, `user_created` hook would send something like:
 
```json
{
  "user": {
    "id": "sqlpad user id",
    "name": "user name"
  } 
}
```

For some background, the REST API used to envelope data like this, but it was removed since it isn't commonly done, and is kind of redundant (you know that you're getting a user if you do GET `/api/users/<id>`)

I wasn't sure if webhooks should play by same rules, or if they should describe the object.

Something that can be done if enveloping is that you can add metadata and other info to the object without it being out of place. Instead of placing the SQLPad public url in a header, it could be something like 

```json
{
  "meta": {
    "sqlpadUrl": "https://yoursqlpadurl.com"
  },
  "user": {
    "id": "sqlpad user id",
    "name": "user name"
  } 
}
```

Thoughts?

Reply:
how about just renamind "id" --> "user_id" ?

Reply:
is that weird if that doesn't align with API object model?

GET /api/user/1 returns `{ id, name, email, role, ...someOtherFields }`, hook returns `{ userId, name, email, role }`

In the query object, there's a`createdByUser` object that has an `id` field as opposed to `userId` as well.

Reply:
I don't think it is too confusing.

However I have no relevant experience nor strong opinion

Reply:
I think there's some benefit to having the same object schema/model passed personally. But I'm only saying that as someone who also anticipates working with the API responses as well.

I've been reading through GitHub's webhook docs, and their webhooks send enveloped payloads allowing more metadata to be passed. Though GitHub has a lot of context to send for various events, like repository for whatever the action pertains to https://developer.github.com/webhooks/event-payloads/#webhook-payload-example-7

Reply:
Reflecting further and I think I'm going the envelope route. Hook payloads can then contain the metadata instead of putting that in headers, which I think makes things a bit more convenient for someone implementing the hook.

The secret would stay in a header, but that can be processed in a middleware, and then ignored onward. 

The hook name and public URL of an instance are relevant pieces of data for doing things with the payload, so it seems fine to add them in the body. 

This makes it easy if the consumer is storing the payload too. If included they can just pass the body on to the storage or service of their choice. If not included, they'd need to pull info out of headers and decorate accordingly anyways.